### ISSUE TITLE
Magit no longer works with cygwin ?

### ISSUE THREAD

**Issue by nchapon** [magit/magit#2246]
It seems magit version 20150907.504 doesn't work with cygwin on windows 

When I try to open magit-buffer with magit-status command :
magit-insert-head-header: Wrong type argument: stringp, nil

It was working fine with previous versions (with magit 2.1.0). Did you change something concerning how git command is running ?

Thanks

**Comment by tarsius** [comment]
> Did you change something concerning how git command is running ?

No.

**Comment by tarsius** [comment]
Try `M-x toggle-debug-on-error`, possibly with all `*.elc` files removed.

**Comment by johnmastro** [comment]
AFAICT it has something to do with the curly brackets in e.g. "^{commit}" not being quoted. I don't yet understand why additional measures appear to be necessary with Cygwin or what the best fix would be, but this seems to help in the meantime:

    (defun magit-git-insert-quote-curlies (args)
      (mapcar (lambda (arg)
                (if (stringp arg)
                    (replace-regexp-in-string "{\\(commit\\|tree\\)}"
                                              "\\\\{\\1\\\\}"
                                              arg)
                  arg))
              args))

    (advice-add 'magit-git-insert
                :filter-args
                #'magit-git-insert-quote-curlies)

i also tried this but it wasn't enough to get stashes working again. This afternoon I'll try just replacing all curly brackets (I was trying not to be too indiscriminate).

    (defun magit-git-insert-quote-curlies (args)
      (mapcar (lambda (arg)
                (if (stringp arg)
                    (thread-last arg
                      (replace-regexp-in-string "{\\(commit\\|tree\\)}"
                                                "\\\\{\\1\\\\}")
                      (replace-regexp-in-string "stash@{\\([0-9]+\\)}"
                                                "stash@\\\\{\\1\\\\}"))
                  arg))
              args)

**Comment by tarsius** [comment]
Ah, that again :-( Good find, I would never have though of it.

**Comment by tarsius** [comment]
We already try to deal with this using this:

```lisp
(defun magit-process-git-arguments (args)
  "Prepare ARGS for a function that invokes Git.

Magit has many specialized functions for running Git; they all
pass arguments through this function before handing them to Git,
to do the following.

* Flatten ARGS, removing nil arguments.
* Prepend `magit-git-global-arguments' to ARGS.
* Quote arguments as required when using Powershell together
  with Cygwin Git.  See #816."
  (setq args (-flatten args))
  (when (and (eq system-type 'windows-nt)
             (let ((case-fold-search t))
               (string-match-p "cygwin" magit-git-executable)))
    (setq args (--map (replace-regexp-in-string
                       "{\\([0-9]+\\)}" "\\\\{\\1\\\\}" it)
                      args)))
  (append magit-git-global-arguments args))
```

What would have to be changed for this to kick in in your case?

**Comment by tarsius** [comment]
Meh, obviously extending the regex might be enough.

**Comment by tarsius** [comment]
Obviously we need to avoid false-positives as much as possible does `"{commit}"` get messed up or only `"^{commit}"`?

**Comment by tarsius** [comment]
What happens if you try to stage a file named `"{file}"`?

**Comment by johnmastro** [comment]
I'll do some experiments and get back to you.

Part of the problem for me specifically appears to be that I use a Cygwin prefix of `/`, so my `magit-git-executable` is `c:/bin/git.exe` and the Cygwin handling isn't triggered.

I'm not sure what the best fix there would be - maybe a `magit-using-cygwin` `defcustom`?

**Comment by tarsius** [comment]
> I'm not sure what the best fix there would be - maybe a `magit-using-cygwin` `defcustom`?

Someone might use cygwin sometimes, sometimes not. Also we would want to guess the default, so we need to figure out whether cygwin is in use, even if we do add an option. Maybe we could check for the availability of some executable.

**Comment by johnmastro** [comment]
Good point. Looks like `(executable-find "cygpath.exe")` will work as the test.

So far handling either `"{commit}"` or just `"^{commit}"` have both allowed me to use `magit-status`, stage and unstage, commit, view the log, and rebase. Do you know offhand something that should fail if handling just `"^{commit}"` isn't enough?

However, with either of the above changes `z z` still fails for me with the following backtrace. I don't know if it's directly related to this issue though. Do you want to address it here as well or should I open a separate issue?

    Debugger entered--Lisp error: (error "Cannot save the current worktree state")
      signal(error ("Cannot save the current worktree state"))
      error("Cannot save the current worktree state")
      magit-stash-create("On master: test-stash" t t nil)
      magit-stash-save("On master: test-stash" t t nil t)
      magit-stash("On master: test-stash" nil)
      funcall-interactively(magit-stash "On master: test-stash" nil)
      #<subr call-interactively>(magit-stash nil nil)
      ad-Advice-call-interactively(#<subr call-interactively> magit-stash)
      apply(ad-Advice-call-interactively #<subr call-interactively> magit-stash)
      call-interactively(magit-stash)
      magit-invoke-popup-action(122)
      funcall-interactively(magit-invoke-popup-action 122)
      #<subr call-interactively>(magit-invoke-popup-action nil nil)
      ad-Advice-call-interactively(#<subr call-interactively> magit-invoke-popup-action nil nil)
      apply(ad-Advice-call-interactively #<subr call-interactively> (magit-invoke-popup-action nil nil))
      call-interactively(magit-invoke-popup-action nil nil)
      command-execute(magit-invoke-popup-action)

**Comment by npostavs** [comment]
> Good point. Looks like (executable-find "cygpath.exe") will work as the test.

But it shouldn't trigger in cases that cygwin is installed but cygwin git is not being used, so maybe

```elisp
(let ((exec-path (file-name-directory magit-git-executable)))
  (executable-find "cygpath.exe"))
```

**Comment by johnmastro** [comment]
> But it shouldn't trigger in cases that cygwin is installed but cygwin git is not being used, so maybe
>
> (let ((exec-path (file-name-directory magit-git-executable)))
>   (executable-find "cygpath.exe"))

Ah, indeed. With that in mind, here's what I'm using currently that appears to work for me, except for the apparently separate issue of stashing.

    (defun magit-process-git-arguments (args)
      (setq args (-flatten args))
      (when (and (eq system-type 'windows-nt)
                 (let ((exec-path
                        (list (file-name-directory magit-git-executable))))
                   (executable-find "cygpath.exe")))
        (setq args (--map (let* ((it (replace-regexp-in-string
                                      "{\\([0-9]+\\)}" "\\\\{\\1\\\\}" it))
                                 (it (replace-regexp-in-string
                                      "\\^{commit}" "^\\\\{commit\\\\}" it)))
                            it)
                          args)))
      (append magit-git-global-arguments args))

**Comment by tarsius** [comment]
But we shouldn't use `executable-find` every time git is called, so we probably need option `magit-process-git-arguments-function` and function `magit-process-git-arguments-cygwin`.

**Comment by johnmastro** [comment]
When should we set `magit-process-git-arguments-function` - when `magit-status` is invoked or less frequently, like when `magit` is first loaded?

**Comment by tarsius** [comment]
When it's first loaded, so the test would move into the `defcustom` form.

Hm, or maybe we should think about it some more. After all, one might also use tramp to connect from cygwin to non-cygwin.

**Comment by tarsius** [comment]
So in any case we should first figure out if cygwin has an option "I am not gonna quote curlys - deal with it".

**Comment by johnmastro** [comment]
I'll look into that today

**Comment by thierryvolpiatto** [comment]
Not sure if it is related, but for me magit is broken also in GNU/Linux
with same symptoms, I had to switch to 2.2 which is working fine.

I will try to switch back ASAP to dev version to provide a backtrace.

-- 
Thierry
Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

**Comment by tarsius** [comment]
What broke? Just the stashing? That's what I am going to look into now. There's already a proposed fix at #2253.

**Comment by bsuh** [comment]
#2253 is a performance fix. Not a bug fix.

I had the same problem yesterday. It is fixed on 5309fe9

**Comment by johnmastro** [comment]
So far I'm having positive results by adding `"CYGWIN=noglob"` to the process environment; it seems to convince Cygwin to leave the curly braces alone.

    (defun magit-git-insert-noglob (original &rest args)
      (let ((process-environment (cons "CYGWIN=noglob" process-environment)))
        (apply original args)))

    (advice-add 'magit-git-insert :around #'magit-git-insert-noglob)

(Obviously we wouldn't use advice for this in Magit, it's just an easy way to experiment).

I still want to research this and experiment with it further (haven't had much time today) but wanted to let you know in the meantime that there is some progress.

Back in 2011 [there were complaints](https://cygwin.com/ml/cygwin/2011-11/msg00225.html) that `"noglob"` is buggy in the presence of double quotes. I haven't had any issues yet but will try to confirm if the bugs have been fixed. (A `"glob:nobrace"` variant that would also work for us was proposed at that time but hasn't been implemented.)

**Comment by johnmastro** [comment]
Or maybe the fix is simply to suggest that people using Cygwin from a native Windows emacs add `    (push "CYGWIN=noglob" process-environment)` to their `init.el`.

**Comment by tarsius** [comment]
https://cygwin.com/cygwin-ug-net/using-cygwinenv.html

> (no)glob - if set, command line arguments containing UNIX-style file wildcard characters (brackets, braces, question mark, asterisk, escaped with \) are expanded into lists of files that match those wildcards. This is applicable only to programs run from non-Cygwin programs such as a CMD prompt. That means that this setting does not affect globbing operations for shells such as bash, sh, tcsh, zsh, etc. Default is set.

(I have omitted the details about `[:ignorecase]`.

So in other words the "CMD prompt" does not support globing but users wanted it to do globing, after all they were using unix commands in a shell and a unix shell does support globing. There would have been two ways to deal with that (a) telling users that if they want a unix shell then they should use a unix shell (b) patch the "CMD prompt". Unfortunately the "CMD prompt" cannot be patched because it is closed source.

So someone had the idea, that every cygwin should check whether the parent process is a non-cygwin and if so do the globing on behalf of that. That's a dirty hack, but I guess that for the "CMD prompt" this makes sense.

Unfortunately however the person who implemented that hack did not consider that not *every* non-cygwin process which calls cygwin executables is the "CMD prompt" (or some other defective shell). What should have been done is to provide some sort of whitelist of "non-cygwin shells that need to be patched by their cygwin child processes".

But that's not what happened and so now every non-cygwin process which invokes a cygwin process has to say "oh and by the way, I am not a defective shell".

In Magit we can probably do that by adding this to `magit-call-process` and `magit-start-process`:

```lisp
(let ((process-environment process-environment))
  (setenv "CYGWIN" (--if-let (getenv "CYGWIN") (concat "noglob " it) "noglob")))
  ...)
```

**Comment by tarsius** [comment]
Please check who wins in `CYGWIN="glob noglob"` and `CYGWIN="noglob glob"`.

**Comment by johnmastro** [comment]
In `CYGWIN="glob noglob"`, `noglob` wins; in `CYGWIN="noglob glob"`, `glob` wins. So it looks like whichever comes last takes precedence.

**Comment by johnmastro** [comment]
> In Magit we can probably do that by adding this to magit-call-process and magit-start-process

I think we may also need to add `magit-process-file` with the same logic and then change e.g. `magit-git-exit-code` and `magit-git-insert` to use it.

I haven't found any issues with this approach, or any better alternatives. If you agree that it's the way to go I'd be happy to put together a PR.

**Comment by jasminpatry** [comment]
This bug also occurs with MSYS2 git, which is what I'm currently using. MSYS2 is a fork of Cygwin, and supports the noglob option via `MSYS=noglob`. Unfortunately there's a bug in the current version of MSYS2 which breaks `noglob`, but I helped the maintainers track it down and they just submitted a fix: https://github.com/Alexpux/MSYS2-packages/commit/c1b58eb8c71b0e026bb82463183a51dce6ab8e8b
I've tested this locally and it works; i.e., magit works again if I use @johnmastro's `noglob` fix in https://github.com/magit/magit/issues/2246#issuecomment-139067687 (replacing CYGWIN with MSYS).

So, if you choose to use the noglob fix, please add noglob to both the CYGWIN *and* MSYS environment variables!

**Comment by johnmastro** [comment]
I had a go at this - [diff here](https://github.com/magit/magit/compare/master...johnmastro:jbm/noglob)

**Comment by johnmastro** [comment]
This is probably overkill, but we could also generalize this a bit and/or make it a no-op by default.

I'm on the tired side and the below is untested by hopefully shows what I mean.

```emacs-lisp
;; nil by default?
(defvar magit-process-environment-additions
  '(("CYGWIN" "noglob")
    ("MSYS"   "noglob")))

(defmacro magit-with-process-environment (&rest body)
  (declare (indent 0))
  (let ((var-sym (make-symbol "var"))
        (val-sym (make-symbol "val")))
    `(let ((process-environment process-environment))
       (pcase-dolist ((`(,,var-sym ,,val-sym)
                       magit-process-environment-additions))
         (setenv ,var-sym (--if-let (getenv ,var-sym)
                              (concat it " " ,val-sym)
                            ,val-sym)))
       ,@body)))

(defun magit-process-file (&rest args)
  "Process files synchronously in a separate process.
Identical to `process-file' but wrap the call with
`magit-with-cygwin-noglob'."
  (magit-with-process-environment (apply #'process-file args)))
```

**Comment by npostavs** [comment]
Actually, it's not safe to use `setenv`, even with let-binding `process-environment`. `setenv` can modify the actual list object so the changes may persist.

**Comment by tarsius** [comment]
@npostavs When would it persist? Whenever an error occurs in BODY?

@johnmastro Looks good. I would probably prefer putting the `let` into each caller and using a function instead of `magit-with-cygwin-noglob`. At first I though `magit-process-environment-additions` was a nice idea, but it would be more complicated than that, because not every environment variable has the same semantics as these two.

**Comment by npostavs** [comment]
> When would it persist? Whenever an error occurs in BODY?

No, it's whenever the environment variable is already set. In this case `setenv` uses `setcar`.

```elisp
(progn 
  (setenv "CYGWIN" "xxx") ; user's existing setting
  (let ((process-environment process-environment))
    (setenv "CYGWIN" (--if-let (getenv "CYGWIN")
                         (concat it " noglob")
                       "noglob")))
  (getenv "CYGWIN")) ; => "xxx noglob", i.e. the noglob persists.
```

**Comment by tarsius** [comment]
Aha! Well then we can just use a copy of the original list, right? I have to admit I have not actually read any code after you said the above, and just assumed that what you said means that `setenv` manipulates some C data based on the value of `process-environment`. Quite silly really.

**Comment by npostavs** [comment]
Or use `push`+`concat`

```elisp
(progn 
  (setenv "CYGWIN" "xxx")
  (let ((process-environment process-environment))
    (push (concat "CYGWIN=" (--if-let (getenv "CYGWIN")
                                (concat it " noglob")
                              "noglob"))
          process-environment))
  (getenv "CYGWIN")) ; => "xxx"
```

**Comment by johnmastro** [comment]
Revised diff [here](https://github.com/magit/magit/compare/master...johnmastro:jbm/noglob) building the environment variables in a function (rather than macro) and appending them to `process-environment`. 

```emacs-lisp
(defun magit-cygwin-noglob-env-vars ()
  (mapcar (lambda (var)
            (concat var "=" (--if-let (getenv var)
                                (concat it " noglob")
                              "noglob")))
          '("CYGWIN" "MSYS")))

(defun magit-process-file (&rest args)
  (let ((process-environment (append (magit-cygwin-noglob-env-vars)
                                     process-environment)))
    (apply #'process-file args)))
```

**Comment by thierryvolpiatto** [comment]
I have reinstalled and all is working fine for me now.
Was probably something unrelated to this, and it have been fixed.

Thanks.

**Comment by npostavs** [comment]
Setting `CYGWIN=noglob` breaks "Git for Windows" (at least with version 1.9.5), so this will have to be configurable.
```
128 ../magit/ c:\Program Files\MSysGit\bin\git.exe … fetch origin
Invalid command: '"git-upload-pack 'npostavs/magit.git'"'
  You appear to be using ssh to clone a git:// URL.
  Make sure your core.gitProxy config option and the
  GIT_PROXY_COMMAND environment variable are NOT set.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

Looks like some shell expansion fails to occur. 

---

I have just now seen that Git for Windows has updated to 2.5. I haven't managed to use it successfully yet, I think due the directory layout being reorganized quite a bit.

**Comment by tarsius** [comment]
I have pushed @johnmastro commits plus some fixup commits to `jm/noglob`.

**Comment by johnmastro** [comment]
In the last commit on `jm/noglob` (d5d6754), `magit-start-file-process` is used on line 430 but the definition has been removed.

**Comment by johnmastro** [comment]
> Setting CYGWIN=noglob breaks "Git for Windows" (at least with version 1.9.5), so this will have to be configurable.

They sure don't make this easy.

Besides making it configuration, another option would be to have `magit-cygwin-noglob-env-vars` use the check we considered before. Something like this:

```emacs-lisp
(defun magit-cygwin-noglob-env-vars ()
  (when (let ((exec-path (list (file-name-directory magit-git-executable))))
          (executable-find "cygpath"))
    (mapcar (lambda (var)
              (concat var "=" (--if-let (getenv var)
                                  (concat it " noglob")
                                "noglob")))
            '("CYGWIN" "MSYS"))))
```

**Comment by npostavs** [comment]
It occurs to me that
```elisp
(let ((exec-path (list (file-name-directory magit-git-executable))))
  (executable-find "cygpath.exe"))
```
is pretty much just `(file-executable-p (concat (file-name-directory magit-git-executable) "cygpath.exe"))`

**Comment by johnmastro** [comment]
Indeed. I edited the code in my comment above to use that instead

**Comment by johnmastro** [comment]
It seems MSYS2 uses `pacman` for package management, so maybe something like `(file-executable-p (concat (file-name-directory magit-git-executable) "pacman.exe"))` would be good enough for detecting MSYS2.

**Comment by npostavs** [comment]
Some good news: Git for Windows 2.5 (probably 2+) works regardless of `noglob` settings.

---

I'm wondering if instead of looking for magic files we just execute something like `git -c alias.echo=!echo echo x{0}` and check that the output is `x{0}`.

**Comment by johnmastro** [comment]
> I'm wondering if instead of looking for magic files we just execute something like git -c alias.echo=!echo echo x{0} and check that the output is x{0}.

That sounds like an excellent idea to me - why look for a proxy indicator when you can test for the actual issue? And I can confirm that it does detect the presence/absence of the issue on the systems I have access to.

I tested with this:

```emacs-lisp

(defun broken-globing-p ()
  (equal "x0\n" (with-temp-buffer
                  (process-file magit-git-executable
                                nil (current-buffer) nil
                                "-c" "alias.echo=!echo" "echo" "x{0}")
                  (buffer-string))))
```

**Comment by johnmastro** [comment]
I created a branch `jm/noglob` on my fork which is based on the `jm/noglob` @tarsius created with two additional commits:

1. Remove the use of `magit-start-file-process` (it is no longer defined)
2. Use the test @npostavs recommended to only use the environment variables when necessary 

I also tweaked the docstring of `magit-cygwin-noglob-env-vars` to mention that the list returned may be `nil`.

Comparison to `master` [here](https://github.com/magit/magit/compare/master...johnmastro:jm/noglob).

**Comment by npostavs** [comment]
Hmm, but https://github.com/magit/magit/issues/2246#issuecomment-138886922

> we shouldn't use `executable-find` every time git is called

goes double for `call-process`. It's especially slow on Windows.

**Comment by johnmastro** [comment]
So maybe a sketch looks like this:

```emacs-lisp
(defvar magit-env-vars-function nil)

(defun magit-cygwin-env-vars ()
  (when (equal "x0\n" (with-temp-buffer
                        (process-file magit-git-executable
                                      nil (current-buffer) nil
                                      "-c" "alias.echo=!echo" "echo" "x{0}")
                        (buffer-string)))
    (mapcar (lambda (var)
              (concat var "=" (--if-let (getenv var)
                                  '(concat it " noglob")
                                "noglob")))
            '("CYGWIN" "MSYS"))))

(defun magit-process-file (&rest args)
  (let ((process-environment (if magit-env-vars-function
                                 (append (funcall magit-env-vars-function)
                                         process-environment)
                               process-environment)))
    (apply #'process-file args)))
```

And then those of use using Cygwin or MSYS2 put this in our `init.el`:

```emacs-lisp
(setq magit-env-vars-function #'magit-cygwin-env-vars)
```

**Comment by johnmastro** [comment]
Or, of course, we could remove the test entirely and teach people to set the variable appropriately. Or we could do something like the below - detect the issue up front, and then people only need to set it if they change what git they're using during a session.

```emacs-lisp
(defvar magit-env-vars-function
  (and (equal "x0\n" (with-temp-buffer
                       (process-file magit-git-executable
                                     nil (current-buffer) nil
                                     "-c" "alias.echo=!echo" "echo" "x{0}")
                       (buffer-string)))
       #'magit-cygwin-env-vars))
```

**Comment by npostavs** [comment]
That still calls a process to check every time, if the option is set. I think we want something a bit more like

```elisp
(defvar magit-need-cygwin-noglob
  (equal "x0\n" (magit-git-string "-c" "alias.echo=!echo" "echo" "x{0}")))

(defun magit-cygwin-env-vars ()
  (when magit-need-cygwin-noglob
    (mapcar (lambda (var)
              (concat var "=" (--if-let (getenv var)
                                  '(concat it " noglob")
                                "noglob")))
            '("CYGWIN" "MSYS"))))

(defun magit-process-file (&rest args)
  (let ((process-environment (append (magit-cygwin-env-vars)
                                     process-environment)))
    (apply #'process-file args)))
```

**Comment by tarsius** [comment]
@npostavs Yep, I think that's what we should do. (Except for the stray `'` of course.)

**Comment by johnmastro** [comment]
I'll do that and submit it as a pull request. Should `magit-need-cygwin-noglob` be a `defvar` or a `defcustom`?

**Comment by tarsius** [comment]
Use an option. I would also drop `magit-cygwin-env-vars`s doc-string, unless you can come up with a summary that fits on one line and is a bit more useful than "Return a list of environment variables." (I can't).

**Comment by johnmastro** [comment]
Sounds good. I've submitted it as pull request #2281 

> I would also drop magit-cygwin-env-varss doc-string, unless you can come up with a summary that fits on one line and is a bit more useful than "Return a list of environment variables." (I can't).

Anything useful ended up too long, so I removed it.

I didn't use `magit-git-string` in the definition of `magit-need-cygwin-noglob` because it's not available yet (because `magit-process-file` isn't defined yet) but that seemed like the most natural place for the option.

How's the commit message?
